### DISCUSSION TITLE
NMEA2000

### DISCUSSION THREAD

**Discussion by peterantypas** [peterantypas/maiana#67] [Ideas]
Early this week, I sent a survey to people who signed up on the mailing list. The purpose of the survey was to determine what type of connectivity MAIANA needs to support to serve as many boaters as possible. Here are the results:

![Screenshot from 2021-06-18 12-10-40](https://user-images.githubusercontent.com/1565933/122606794-5d75b400-d02e-11eb-831d-15da4c0b3496.png)

As you can see, NMEA2000 slightly edges NMEA0183. Therefore, I started a discussion with Timo Lappalainen who authored this excellent [NMEA2000](https://github.com/ttlappalainen/NMEA2000) package, and I think I have a viable path to get to a working N2K interface reasonably soon, using the exact same MCU I have stocked for the transponder (which doesn't have a CAN peripheral BTW). This means that the MAIANA control box / breakout board will eventually have a NMEA2000 interface and publish AIS messages to that bus.

To take things a bit further (because that's my nature), I am thinking of building this as a standalone module that can be easily dropped into other designs and provide bridging functionality. Think of what cellular and BLE modules have done for IoT development, and imagine having a NMEA2000 module with a nice, clean&open SDK to help integrate myriads of devices to that bus without forcing them to use clumsy Arduino, RPis or even Teensy boards. Timo has done an incredible job bringing this functionality to these platforms, but I think the ultimate coup against the NMEA will be to make this an open, drop-in module with a simple command interface that even beginners can understand and use.

If the complexity of an HTTPS transaction can be reduced to AT commands (see any modern cellular modem), then N2K is definitely doable. 

Let me know if you're interested in participating. I'd like this to be more of a community project. I haven't fleshed out any details, I just want to plant this idea out there ;)

**Comment by kagouraki** [comment]
You can count me in. I will help as much as i can. 
George

**Comment by astuder** [comment]
A generic NMEA2000 module would definitely be of interest for my own projects.

**Reply by peterantypas** [reply]
Just for kicks, I fired off an email to NMEA to see if such a device can earn certification. My thinking is that if they're receptive to it, then it makes sense to join NMEA and run a Kickstarter campaign to make it happen. I don't expect a response from them, of course, but there's no downside to asking anyway.

On another note, I took a stab at the hardware design and looks like it can fit in a 19x19 mm LGA form factor with a 3.3mm height (including can). Should be small enough, I hope?

![n2kmodule-1 0-top](https://user-images.githubusercontent.com/1565933/122651352-1ac6e100-d0ed-11eb-819c-7bf766e5726f.png)

![n2kmodule-1 0-bottom](https://user-images.githubusercontent.com/1565933/122651360-1ef2fe80-d0ed-11eb-932f-88049979e065.png)

In terms of connections, I'm thinking 12V, GND, CAN_H, CAN_L, a UART at >= 230kbps for the main interface, maybe an I2C slave as well for configuration, a 5V output from the internal SMPS, possibly a 3.3V regulated output as well. The "dev kit" would have this module mounted onto a breakout board with breadboard compatible pin headers plus a NMEA 2000 connector. It could also serve as a reference design for integration.

**Reply by astuder** [reply]
Size looks great. Any particular reason to go that fast on the UART side? The more common 115200 should be plenty for most if not all NMEA 2000 applications.

**Reply by peterantypas** [reply]
The nominal bus speed is 250kbps, but now that you mentioned it I think it's fine to default to 115200 and have a command for resetting the baud rate to something higher just like modems do.

**Reply by astuder** [reply]
Integrated 12V to 5V DC/DC converter would be great.

Did you also consider extra ESD/surge protection on the CAN and power pins? Or do you expect that to be added externally between the module and N2K connector?

**Reply by peterantypas** [reply]
I think those may need to be external, otherwise I'll have to grow the module. Most modems I've designed into circuits also expect such devices to be external.

**Reply by peterantypas** [reply]
But maybe it makes sense to grow the module just to include them ... how big would be too big you think?

**Reply by astuder** [reply]
I can see why that's expected to be external. It would be hard to spec a protection that makes everyone happy without being overkill 99% of the time.

Module size: I was to say about the size of an ESP8266 module, but I just measured one and you're already at that scale (12x20mm). :)

**Reply by peterantypas** [reply]
I'm looking at some microscopic TI switches to help shrink the SMPS. This is the biggest offender right now (it's the entire northwest region on the layout)

**Reply by muebau** [reply]
Would the CAN bus isolated like transceivers as the "ISO1050" does?

It could help with galvanic corrosion in situations when you connect to parts like the sensors of the engine etc. .

**Reply by peterantypas** [reply]
Good point. It won't fit in this form factor, for sure. Also, what about power in this case? Don't you need isolation there as well?

**Reply by peterantypas** [reply]
To operate in isolation, the transponder will need a 12V-12V buffer like these: https://www.mouser.com/datasheet/2/281/KDC_MEE3C.A01-1186156.pdf

These add significant cost, plus overhead of extra filtering components. Are there any standard NMEA2000 isolators that can be inserted between the bus and the device?

**Reply by muebau** [reply]
Yeah I think a isolation (if needed) would need an isolated power supply too. The batteries minus is connected to the engine anyway the isolation is useless I guess. It was just a idea but I think there were missing parts in it.

**Reply by peterantypas** [reply]
It's easy to miss that part. When I converted my boat to electric propulsion ten years ago, the motor manufacturer strongly urged me to install a galvanic isolator for mains ground, since the battery bank is charged via shore power. I'm glad I did :)

**Comment by peterantypas** [comment]
Now that I'm starting to lay down the framework for N2K, I'm thinking of getting an iKonvert from Digital Yacht for testing: https://digitalyachtamerica.com/product/ikonvert-usb/

I already have their N2K backbone, so ideally I should be able to use this to validate MAIANA's NMEA2000 implementation. Are there any other devices that I should consider instead / in addition to?

**Comment by gah67** [comment]
Isn't the iKonvert just a really expensive 'marine' version of something like https://amazon.co.uk/Q-BAIHE-Analyzer-Transparent-Transmission-Converter/dp/B017N7YEAO with a NMEA2000 plug?

**Reply by peterantypas** [reply]
I don't think so. NMEA2000 is not just vanilla CAN. Unless a device has full support for N2K "fast packet" fragmentation and reassembly, it is not N2K compliant. Most plain CAN implementations do not, and that's why Timo has written replacement libraries for a bunch of them:

https://github.com/ttlappalainen/NMEA2000

But I can't strictly rely on library X to confirm compatibility between X and standard Y. I need at least one certified specimen of Y. That's why I'm leaning toward iKonvert. But I'm not necessarily sold on it yet.

**Reply by muebau** [reply]
As I plan to run nearly 100% with ttlappalainen/NMEA2000 its good enough for me. The network is based on OpenPlotter and several "ttlappalainen" nodes. If there would be an anemometer somewhere in some time it would work in this setup too.

I thought to buy the NMEA0183 version and put a NMEA0183<->NMEA2000 (ttlappalainen) in front to have the best of the two worlds. This way it should be fine.

The "decision" of NMEA0183/NMEA2000/RS232-TTL/USB-Serial is done in the connection box completely, right? So it should be possible to replace it at some time if a different connection type would be needed and a change would be worth the overhead.

**Reply by peterantypas** [reply]
You're a special case. Since you're running on library X and I'm building the adapter with library X, things should be fine. But many boats just use Y, and that's the risk I have to mitigate.

**Comment by peterantypas** [comment]
The work has begun. I can confirm the adapter is connecting to the bus and claiming an address, but my other NMEA2000 device hasn't arrived yet, so that's when the real fun begins.

![IMG_1209](https://user-images.githubusercontent.com/1565933/124694429-17c24380-de96-11eb-832c-8aaf0732d319.JPG)

**Comment by peterantypas** [comment]
My initial implementation sends PGNs 129038, 129039, 12979, 129809, 129810, 129041. These are strictly AIS. Now, given that MAIANA has its own GNSS, does it make sense to also send PGN 129025 (lat/long)? Or is it likely to confuse other devices if there's an existing GPS on the bus?

**Reply by kagouraki** [reply]
I think it should output everything. Also cog and sog. If someome wants he can filter what he want.

**Reply by kagouraki** [reply]
Or someone can disable the chartplotter internal gps and work with the maianas gnss. If something fails he can turn the build in gps on again.

**Reply by peterantypas** [reply]
Is it necessary to register on the bus as a GNSS device? Currently, it registers as AIS (NMEA2000 type 195 according to https://www.nmea.org/Assets/20120726%20nmea%202000%20class%20&%20function%20codes%20v%202.00.pdf) 

Should it also register a second device as class 145?

**Comment by Rudder2** [comment]
I would like the GPS to output on NMEA2000 bus. I originally was going to
make Maiana my ships GPS but ended up with a UBlox since your project went
on hold but would still like the option to use Maiana's GPS.

**Reply by peterantypas** [reply]
It turns out there is no direct equivalent of the NMEA0183 RMC sentence in NMEA2000. The closest PGN is 129029 which does not include COG and SOG. In fact, it needs the GGA sentence for most of its content. I'm assuming COG and SOG are supposed to be transmitted via PGNs 129025 and 129026. This is going to be a little more complicated than I thought.

**Comment by peterantypas** [comment]
I've had some initial success with the Digital Yacht iKonvert, but there is a bug somewhere with the payload so some data like SOG is wrong. But at least they're communicating, so that's a good start.

**Comment by peterantypas** [comment]
Alright, class A&B position reports work, and so does GNSS, but with one caveat: I need the ability to calculate leap seconds since 1970 in order to generate the proper timestamp for PGN 129029.

I have yet to verify static reports (5 & 24A/B). That's next.

**Comment by gah67** [comment]
Sounds like you're making great progress!!

**Reply by peterantypas** [reply]
Yeah, so far it's going better than I expected. Another PGN I want to support is 129041 (ATON report). Timo's library does not support it, but I think I can take advantage of the iKonvert to reverse engineer it.

**Comment by peterantypas** [comment]
I have successully reverse engineered PGN 129041 so MAIANA will translate AIS message 21 to that. This is helpful, because shore stations transmit navigational aids this way and they're overlaid on the chart. Some of those "aids" are purely virtual, and that's what makes this feature important from a safety perspective.

**Reply by muebau** [reply]
A very cool feature!

**Comment by peterantypas** [comment]
Class B position and static reports (messages 18, 24A and 24B) now also map correctly to NMEA 2000. I think the bulk of the work is done.

**Comment by peterantypas** [comment]
I ran a stress test with MAIANA sending NMEA2000 messages to iKonvert which translated them to NMEA0183 and routed them to OpenCPN. Everything worked pretty much as expected, with some minor quirks that shouldn't affect navigation.

Someone is lending me an Actisense gateway device to repeat this test. If all goes well, this will be a wrap by the end of this week.

My NMEA2000 circuit drew less than 10mA from 3.3V during that exercise. This means that a NMEA2000 module based on my architecture will not need an SMPS, just a 2mmx2mm LDO like the one MAIANA already uses. This will shrink it even more.

**Comment by peterantypas** [comment]
Just tested with Actisense NGW-1 and only found one discrepancy: It does not output the name of an AtoN from PGN 129041 into NMEA0183. This is the last field in the packet, so either it's a bug in their firmware, or there's something wrong with my reverse engineering of the packet. It's not a showstopper though.

**Comment by peterantypas** [comment]
Initial board and enclosure fit well:

![IMG_1220](https://user-images.githubusercontent.com/1565933/126045643-9da521c3-112e-4fff-8503-5af3109b0d77.JPG)

That said, this board has 2 major logic issues that I can't fix without a re-spin, so that's in progress.

**Reply by peterantypas** [reply]
One of the issues is #72, which I must absolutely address immediately.

**Reply by peterantypas** [reply]
Alright, #72 has been dealt with, and the board works as expected under normal circumstances.

DFU logic is broken, though, and that's what I need to resolve.

**Comment by peterantypas** [comment]
The final NMEA2000 board spin is now in progress and it should address issue #75. I expect it here on Monday.

**Comment by tdomider** [comment]
I've been using the MAIANA AIS USB + CAN Adapter for years.
After recent experiments, CAN communication stopped working. USB communication works fine.
My board is version 7.0.
I need help repairing it.
